### ISSUE TITLE
`FTB001` should be ignored for positional-only args

### ISSUE THREAD

**Issue by dosisod** [astral-sh/ruff#24408]
### Summary

The following code ([playground link](https://play.ruff.rs/29dfa765-02d8-4641-bda9-6da45951a108)):

```python
def pos_only_arg(x: bool, /) -> None:
    pass


def pos_only_arg_default(x: bool = True, /) -> None:
    pass


pos_only_arg(True)
pos_only_arg(x=True)  # invalid

pos_only_arg_default(True)
pos_only_arg_default(x=True)  # invalid
```

Emits the following errors `FBT` errors:

```
FBT001 Boolean-typed positional argument in function definition
 --> x.py:1:18
  |
1 | def pos_only_arg(x: bool, /) -> None:
  |                  ^
2 |     pass
  |

FBT001 Boolean-typed positional argument in function definition
 --> x.py:5:26
  |
5 | def pos_only_arg_default(x: bool = True, /) -> None:
  |                          ^
6 |     pass
  |

FBT002 Boolean default positional argument in function definition
 --> x.py:5:26
  |
5 | def pos_only_arg_default(x: bool = True, /) -> None:
  |                          ^
6 |     pass
  |

FBT003 Boolean positional value in function call
  --> x.py:9:14
   |
 9 | pos_only_arg(True)
   |              ^^^^
10 | pos_only_arg(x=True)  # invalid
   |

FBT003 Boolean positional value in function call
  --> x.py:12:22
   |
10 | pos_only_arg(x=True)  # invalid
11 |
12 | pos_only_arg_default(True)
   |                      ^^^^
13 | pos_only_arg_default(x=True)  # invalid
   |

Found 5 errors.
```

**Expected Behavior**

* `FBT001`/`FBT002` shouldn't be emitted on positional-only function arguments
* `FBT003` shouldn't be emitted for arguments which are positional-only

**Rationale**

The positional-only keyword is an explicit API design choice by the author and conveys that the positional boolean arguments are desired. Secondly, applying the `FBT003` suggestion would result in a compilation error.

An [example](https://github.com/pwoolvett/flake8_boolean_trap/issues/7) I provided on the `flake-bool-trap` repo is the following:

```python
class BooleanExpression:
    def __init__(self, value: bool, /) -> None:  # notice the `/`
        self.value = value

BooleanExpression(True)
```

As you can see, the use of a boolean positional argument makes sense here.

Separately, Ruff should probably emit a syntax error for `pos_only_arg(x=True)`, as it is invalid at compile time, and trivially detectable while linting.

**Alternatives**

You *could* just suppress `FBT`, but I think explicit usage of `/` should implicitly suppress `FBT` in these scenarios.

### Version

ruff 0.15.4

**Comment by Seinarukiro2** [comment]
Submitted a PR for FBT001/FBT002: #24415

FBT003 is not addressed there — the linter doesn't resolve callee signatures, so it can't determine whether a call-site argument corresponds to a positional-only parameter.

**Comment by MichaReiser** [comment]
> The positional-only keyword is an explicit API design choice by the author and conveys that the positional boolean arguments are desired.

I agree with the first part, but I'm not fully convinced it implies the second. 

For example, the now new false-negative in the Zulip repository is something that I would want to catch if I enable `FBT001`

https://github.com/zulip/zulip/blob/4c1dcd6a7743da817912cc45735a8eae8a35c5d4/zerver/views/auth.py#L656-L670

> Separately, Ruff should probably emit a syntax error for pos_only_arg(x=True), as it is invalid at compile time, and trivially detectable while linting.


This isn't trivial if the function definition lives in a separate file. It then requires multifile analysis, which Ruff doesn't support today.

**Comment by dosisod** [comment]
> For example, the now new false-negative in the Zulip repository is something that I would want to catch if I enable `FBT001`
> 
> https://github.com/zulip/zulip/blob/4c1dcd6a7743da817912cc45735a8eae8a35c5d4/zerver/views/auth.py#L656-L670

Why would reporting `FBT001` make sense for line 660? According to the docs for `FBT001`, you can only resolve the error by:
* Moving `is_signup` after positional only indicator: Allows `is_signup` to be processed as kwarg, which would break compatibility as per the comment, hence why it is before the `/`. The proposed PR should emit `FBT001` if the bool argument is *after* the `/`, but not *before* it.
* Moving `is_signup` after kwarg only indicator: Breaks ABI, as `is_signup` can no longer be used as a positional arg.
* Suppressing `FBT001` via `noqa`: The proposed PR would do just that, as the boolean arg is before the `/`.

I guess the sticking point is whether putting boolean arguments before a `/` should be considered as:
1. An intentional design choice, or
2. An accident, which would require a (potentially) breaking ABI change to fix

**Comment by dosisod** [comment]
Another potential resolution would be introducing a new FBT004 rule which bans the use of boolean args before `/`. 

Assuming the proposed PR is in place, FBT004 would ban positional args before `/`, and FBT001/FBT002 would ban positional bool usage after `/`, or when `/` isn't specified. Then you could disable FBT004 project wide, which allow positional bools before `/`, and warn about positional bools after `/`.

**Comment by MichaReiser** [comment]
> Moving is_signup after kwarg only indicator: Breaks ABI, as is_signup can no longer be used as a positional arg.

I think that's very much the intention of the rule: `Or, make the argument a keyword-only argument:`

> I guess the sticking point is whether putting boolean arguments before a / should be considered as:

I don't think this warrants adding another rule. In fact, it would have to be two extra rules, one for `FBT001` and `FBT002` (the split already feels more like an implementation detail to me).

How common is this pattern in your code? I feel like this is a perfect use case for a noqa, which explicitly documents whether using the bool argument here was an intentional design decision.

**Comment by dosisod** [comment]
I would say this is an uncommon pattern. The source code in question isn't open source, so here is a general example of what it is doing, and where I am encountering the FBT warnings.

```python
from dataclasses import dataclass


@dataclass
class Obj:
    some_very_long_flag_name: bool
    # other unrelated fields


def update_config(filename: str, some_very_long_flag_name: bool) -> None:
    with open(filename, "w+") as f:
        f.write("enabled" if some_very_long_flag_name else "disabled")



def usage() -> None:
    o = Obj(some_very_long_flag_name=True)

    update_config("file.dat", o.some_very_long_flag_name)
    # Here, the field name matches the arg name, which is descriptive enough, but emits FBT001


    # Alternatives

    update_config("file.dat", o.some_very_long_flag_name=)
    # PEP-736 shorthand: PEP was rejected :(

    update_config("file.dat", some_very_long_flag_name=o.some_very_long_flag_name)
    # Explicit kwargs: This gets less tenable as flags are added

    update_config("file.dat", o)
    # Pass config object: workable, but creates a dependency on `Obj`, and the naming of the fields
```

Hopefully that helps. I'm more inclined to disable FBT globally than suppress it at the caller/callee site: The edge cases where I run into the above example outnumber the places where I would like it to warn about "improper" usage.

**Comment by paveldikov** [comment]
> I think that's very much the intention of the rule: `Or, make the argument a keyword-only argument:`

As a spectator, I think this part of the guidance should be made more obvious.

I myself spent an embarrassing amount amount of time figuring out how to reconcile `FBT` with any sort of boolean args. I eventually figured it out & made them kw-only -- and I am now living the happy-ever-after. But it took way too long for the penny to drop -- I'd have just blindly `noqa`d if I weren't such a pedant :-)
